File: group1/102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# class Solution:
#     def levelOrder(self, root: TreeNode) -> List[List[int]]:
#         res = []
#         if not root:
#             return res

#         queue = [root]
#         while(queue):
#             temp = []
#             for _ in range(len(queue)):
#                 p = queue.pop(0)
#                 temp.append(p.val)
#                 if p.left:
#                     queue.append(p.left)
#                 if p.right:
#                     queue.append(p.right)
#             res.append(temp)

#         return res

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        res = []
        if not root:
            return res

        def dfs(idx, root):
            if idx > len(res):
                res.append([])
            left, right = root.left, root.right
            res[idx-1].append(root.val)
            if left:
                dfs(idx+1, left)
            if right:
                dfs(idx+1, right)

        dfs(1, root)
        return res

File: leetcode/花花酱题集/102.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  if (!root) return [];

  const res = [];
  const queue = [root];

  while (queue.length > 0) {
    const n = queue.length;
    const temp = [];
    for (let i = 0; i < n; i++) {
      const { left, right, val } = queue.shift();
      temp.push(val);
      if (left) queue.push(left);
      if (right) queue.push(right);
    }
    res.push(temp);
  }
  return res;
};
